src/main/scala/cromwell/engine/WorkflowActor.scala

```diff
@@ -1,79 +1,85 @@
 import cromwell.engine.backend.Backend
-
+import WorkflowActor._
 import scala.concurrent.ExecutionContext.Implicits.global
 import scala.concurrent.duration._
 import scala.language.postfixOps
-import scala.util.{Failure, Success}
-
 
 object WorkflowActor {
-
   sealed trait WorkflowActorMessage
-
+  case object Complete extends WorkflowActorMessage
   case object Start extends WorkflowActorMessage
-  case object Started extends WorkflowActorMessage
-  case class Done(outputs: Map[String, WdlValue]) extends WorkflowActorMessage
   case class Failed(failure: String) extends WorkflowActorMessage
   case object GetOutputs extends WorkflowActorMessage
+  case object GetFailureMessage extends WorkflowActorMessage
+  
 
-  def props(id: WorkflowId, binding: WdlBinding, coercedInputs: WorkflowCoercedInputs, backend: Backend) =
+  def props(id: WorkflowId, binding: WdlBinding, coercedInputs: WorkflowCoercedInputs, backend: Backend) = {
     Props(new WorkflowActor(id, binding, coercedInputs, backend))
+  }
 
   implicit val ActorTimeout = Timeout(5 seconds)
-}
 
-/**
- * Actor to manage the execution of a single workflow.
- */
-case class WorkflowActor private(id: WorkflowId,
-                                 binding: WdlBinding,
-                                 actualInputs: WorkflowCoercedInputs,
-                                 backend: Backend) extends Actor {
+  sealed trait WorkflowFailure
+  case object NoFailureMessage extends WorkflowFailure
+  case class FailureMessage(msg: String) extends WorkflowFailure with WorkflowActorMessage
+}
 
-  private val log = Logging(context.system, this)
+case class WorkflowActor(id: WorkflowId,
+                         binding: WdlBinding,
+                         actualInputs: WorkflowCoercedInputs,
+                         backend: Backend) extends LoggingFSM[WorkflowState, WorkflowFailure] {
+  implicit val timeout = Timeout(5 seconds)
 
   private val storeActor = context.actorOf(StoreActor.props(binding, actualInputs))
 
-  def receive: Receive = unstarted
+  startWith(WorkflowSubmitted, NoFailureMessage)
 
-  /** Unstarted `Receive` state that only responds to `Start` messages. */
-  def unstarted: Receive = LoggingReceive {
-    case WorkflowActor.Start => handleInitialStart()
-    case bad => fail(bad, "unstarted")
+  when(WorkflowSubmitted) {
+    case Event(WorkflowActor.Start, _) => goto(WorkflowRunning)
   }
 
-  /** Started `Receive` state. */
-  def started: Receive = LoggingReceive {
-    case CallActor.Started(call) =>
+  when(WorkflowRunning) {
+    case Event(CallActor.Started(call), _) =>
       storeActor ! UpdateStatus(call, ExecutionStatus.Running)
-
-    case CallActor.Completed(completedCall, callOutputs) =>
-      handleCallCompleted(completedCall, callOutputs)
-
-    case CallActor.Failed(call, failure) =>
+      stay()
+    case Event(CallActor.Completed(completedCall, callOutputs), _) =>
+      storeActor ! CallCompleted(completedCall, callOutputs)
+      startRunnableCalls()
```

Comment:
I don't think this is right, `startRunnableCalls` should not be async to the completion of `storeActor ! CallCompleted(completedCall, callOutputs)`